force_wfmanager/central_pane/tests/test_analysis_model.py

```diff
@@ -82,3 +83,67 @@ def test_clear(self):
         self.assertEqual(self.analysis.value_names, ())
         self.assertEqual(self.analysis.evaluation_steps, [])
         self.assertEqual(self.analysis.selected_step_indices, None)
+
+    def test_as_json(self):
+        self.analysis.value_names = ('foo', 'bar', 'label')
+        self.analysis.add_evaluation_step((1, 2, 'x'))
+        self.analysis.add_evaluation_step((3, 4, 'y'))
+        self.analysis.add_evaluation_step((5, 6, 'z'))
+        json_rep = {'foo': [1, 3, 5],
+                    'bar': [2, 4, 6],
+                    'label': ['x', 'y', 'z']}
+
+        self.assertEqual(self.analysis.as_json(), json_rep)
+
+    def test_from_dict(self):
+        json_rep = {'foo': [1, 3, 5],
+                    'bar': [2, 4, 6],
+                    'label': ['x', 'y', 'z']}
+        self.analysis.value_names = ('notfoo', 'notbar', 'notlabel')
+
+        self.analysis.add_evaluation_step((11, 22, 'xx'))
+        self.analysis.add_evaluation_step((55, 66, 'zz'))
+
+        self.analysis.from_dict(json_rep)
+        self.assertEqual(self.analysis.value_names, ('foo', 'bar', 'label'))
+        self.assertEqual(len(self.analysis.evaluation_steps), 3)
+        self.assertEqual(self.analysis.evaluation_steps[0], (1, 2, 'x'))
+        self.assertEqual(self.analysis.evaluation_steps[1], (3, 4, 'y'))
+        self.assertEqual(self.analysis.evaluation_steps[2], (5, 6, 'z'))
+
+    def test_write_to_json(self):
+        json_rep = {'foo': [1, 3, 5],
+                    'bar': [2, 4, 6],
+                    'label': ['x', 'y', 'z']}
+        self.analysis.from_dict(json_rep)
+        self.assertEqual(self.analysis.as_json(), json_rep)
+        mock_open = mock.mock_open()
+        with mock.patch('__main__.open', mock_open, create=True):
+            with mock_open('blah', 'w') as fp:
+                success = self.analysis.write_to_json(fp)
+
+        self.assertTrue(success)
+
+    def test_write_to_csv(self):
+        json_rep = {'foo': [1, 3, 5],
+                    'bar': [2, 4, 6],
+                    'label': ['x', 'y', 'z']}
+        self.analysis.from_dict(json_rep)
+        self.assertEqual(self.analysis.as_json(), json_rep)
+        mock_open = mock.mock_open()
```

Comment:
neat I didn't know that mock had this! It also made me very confused since I looked at L139 onwards first

Reply:
It's quite neat! Unfortunately `json.dump` doesn't write line-by-line so I can't do the write call comparison as in the csv example. Probably overkill anyway, and `assert_any_call` doesn't let you enforce an order!